### PR TITLE
Feature/29 jumps

### PR DESCRIPTION
**Pull request by teggotic**
fixes https://github.com/jemc/mare/issues/29
fixes https://github.com/jemc/mare/issues/55

### CODE REVIEW

File: spec/compiler/completeness_spec.cr [hunk 1]
```diff
@@ -32,6 +32,58 @@ describe Mare::Compiler::Completeness do
...
+    expected = <<-MSG
+    This field may be read before it is initialized by a constructor:
+    from (example):2:
+      :prop w U64
+            ^
+
+    - traced from a call here:
+      from (example):5:
+        @w = 0
+        ^~~~~~
```
Comment (jemc) [inline spec/compiler/completeness_spec.cr:54]: This error message is intended to represent that the field may be read during the constructor before it is first initialized, so it doesn't make very much sense to use it for this case (where the field is never read - it is only written, but conditionally so).

I would expect the error to look like this:
```suggestion
    This constructor doesn't initialize all of its fields:
    from (example):3:
      :new
       ^~~

    - this field didn't get initialized:
      from (example):2:
      :prop w U64
            ^

    - it is initialized here, but this expression is not guaranteed to execute:
      from (example):5:
        @w = 0
        ^~~~~~
```

File: spec/compiler/completeness_spec.cr [hunk 2]
```diff
@@ -32,6 +32,58 @@ describe Mare::Compiler::Completeness do
...
```
Comment (jemc) [inline spec/compiler/completeness_spec.cr:79]: Same comment as above.

File: spec/compiler/completeness_spec.cr (hunk 1 above)
Comment (teggotic) [reply spec/compiler/completeness_spec.cr:54]: I don't know why this is resolved as FieldRead
=

File: spec/compiler/completeness_spec.cr [hunk 3]
```diff
@@ -32,6 +32,58 @@ describe Mare::Compiler::Completeness do
...
+      from (example):2:
+      :prop w U64
+            ^
+    MSG
+
+    expect_raises Mare::Error, expected do
+      Mare::Compiler.compile([source], :completeness)
+    end
+  end
+  it "complains when field write is away in constructor" do
```
Comment (jemc) [inline spec/compiler/completeness_spec.cr:61]: Small quibble, but an extra empty line should be between each pair of test cases.

```suggestion

  it "complains when field write is away in constructor" do
```

**Comment by jemc** [comment]
Closing this PR because we will take a new direction for control flow analysis: #75

**Pull request closed without merging**